### ISSUE TITLE
Defining our code review rules

### ISSUE THREAD

**Issue by yaseminalpay** [SWE574-RenameMe/swe574-CoolName#3]
We need to specify our own rules for code review and add it to the wiki page. Let's discuss.

**Comment by muatik** [comment]
There are various open-sourced code reviewing guidelines which we can adopt one of them. If there is any statement that we do not want to follow, we can simply remove it.

1. https://github.com/thoughtbot/guides/tree/master/code-review
2. http://www.slideshare.net/lalitkale/code-review-guidelines
3. https://msdn.microsoft.com/en-us/library/ms182019(v=vs.100).aspx
4. http://matthewmachuga.com/blog/2015/code-review-processes.html

In short, we can choose the first one and make necessary changes to make it suitable for our project.

**Comment by muatik** [comment]
You can have a look at the following url, Python Django framework's main repository url, to see an example of code reviewing on github: https://github.com/django/django/pull/7994/files and https://github.com/django/django/pull/7994/files#diff-ea5b3f195703d9fa27adb7113e7d9c92R88

**Comment by sanal-cem** [comment]
We would start to exercise some rules. These standards vary between different projects. As far as I learned from my previous studies, these rules below should be practiced;
1. Standard variable names should exist.
2. If Else, Switch, Try{} Catch{} writing techniques should not vary from member to member.
3. Average, minimum and maximum number of lines for each function should be decided.
4. SQL Query writing techniques should not vary from member to member. (Different JOIN styles would occur, very expert SQL Queries would not understood, etc.)
5. How many words should be placed in a line? (Mostly for SQL Queries)
6. How to do object transmissions between page to page? eg. from HTML to Java.

Furthermore, as all know, seperating techniques for classes, packages; eg. MVC and rules for objects like getters and setters should exist, etc. should be practiced.

**Comment by muatik** [comment]
@sanal-cem I think we are not discussing which technology standards we should follow. Rather, we need to define a guideline about how we are going to evaluate each other work. Let's say you are implementing the issue 78. When you feel that you finish it, you are supposed to send a pull request stating what you have done, how and why, if necessary. Then, somebody should review your work to prevent bugs earlier, to ensure code quality at a defined level. In addition, the reviewers can also gain more knowledge about all over the project's implementation.

So, we need to define the way, the language, I guess. When you make a wrong design decision, for example, which language should I use, should I be **polite** or what? When I find a defect in your work, **should I immediately reject the pull request**, or if you insist that what you have done is quite right, should I accept it and merge it into the main code repository? If you did not write enough or proper unit tests for your implementation, should I reject your work? These are the things we need to define. **Please follow the links I shared above.** 

We need to define everything each reviewer is going to check. Coding standards, unit and integration tests, proper inline comments, possible and quick improvement suggestions, possible bugs, whether decided git flow is being followed or not, etc... Most of these depend on tech stack. If you implement something in java for android, for instance, then coding/naming standards should be specialized for this environment.

**Comment by fkucuk** [comment]
To say the truth, this will be first for me. I mean Code Reviewing. 
I've read all the documents shared on this issue. 
The number 1 seems suitable for our project. But as the other sources recommended there should be a checklist which the reviewer will follow.
And on the slideshare link there is a suggestion about assigning severity to review findings. Which reminds the reviewer to focus on functional issues before naming standards for instance.

![image](https://cloud.githubusercontent.com/assets/3140657/23095761/07713150-f618-11e6-843f-55c2608d8668.png)

**Comment by muatik** [comment]
http://blog.felipe.rs/2017/02/25/id-software-programming-principles/

**Comment by muatik** [comment]
Is everybody ok with https://github.com/thoughtbot/guides/tree/master/code-review ? If you have any objection, please share it with us. If you are ok, send a comment with a 👍  thumb up. When we reach 8 thumbs up, then we can close this.

**Comment by muatik** [comment]
👍

**Comment by sanal-cem** [comment]
👍

**Comment by yaseminalpay** [comment]
👍

**Comment by tokonu** [comment]
👍

**Comment by kartalerkoc** [comment]
👍

**Comment by sanal-cem** [comment]
👍

**Comment by fkucuk** [comment]
Thumb up ;)
